In [4]:
import pandas as pd
import numpy as np

In [5]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [6]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There are a number of issues with the data as it currently exists. Firstly, the column titles are not standardised(using camelcase rather than snakecase). They are internally consistent, but it would nonetheless be nice to alter them to make the data easier to manage. Secondly, we'll need to be aware of the null values in the data moving forward. Thirdly, some columns are types which are not ideal, such as price and odometer being objects rather than integers or floats. It would be  nice to fix this. Finally, the text is in German, which is not an issue that necessarily needs fixing, although we will have to be aware of German characters present in strings that are not in the English language.

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
modified_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen']
autos.columns = modified_columns

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The first issue outlined above has been fixed; column names have been changed to snakecase and some names have been slightly altered.

In [10]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Titles are looking better, but we still need to fix the price and odometer columns so that they're numeric values rather than text.

In [11]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(float)
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(float)
autos = autos.rename(index = str, columns = {"odometer": "odometer_km"})

In [14]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [15]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [16]:
autos["odometer_km"].value_counts(ascending = False).head(5)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64

In [17]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

There are some clear outliers here. The minimum of 0, and the maximum of 10 million are obviously unusual, so let's look a little closer at the data around these regions.

In [21]:
autos_under_100 = autos[autos["price"] < 100]
autos_under_100["price"].describe()

count    1762.000000
mean        5.632804
std        18.374263
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max        99.000000
Name: price, dtype: float64

So out of 50,000 sales, a little over 3% are below 100. We can reasonably cut these.

In [22]:
autos = autos[autos["price"] >= 100]
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  4.823800e+04       48238.000000  48238.000000   48238.000000   
mean   1.019927e+04        2004.728534    117.692483  125915.149882   
std    4.898087e+05          87.885189    201.207082   39547.782619   
min    1.000000e+02        1000.000000      0.000000    5000.000000   
25%    1.250000e+03        1999.000000     73.000000  125000.000000   
50%    3.000000e+03        2004.000000    107.000000  150000.000000   
75%    7.499000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  number_of_pictures   postal_code  
count        48238.000000             48238.0  48238.000000  
mean             5.801236                 0.0  50988.722045  
std              3.677147                 0.0  25739.558428  
min              0.000000                 0.0   1067.000000  
25%              3.000000                 0.0  30823.000000  
50%              6.000000                 0.0  49716.000000  
75%              9.000000                 0.0  71672.000000  
max             12.000000                 0.0  99998.000000

What about outliers on the high end for prices?

In [23]:
autos_above_100000 = autos[autos["price"] > 100000]
autos_above_100000["price"].describe()

count    5.300000e+01
mean     4.013343e+06
std      1.435784e+07
min      1.049000e+05
25%      1.290000e+05
50%      1.800000e+05
75%      9.999900e+05
max      1.000000e+08
Name: price, dtype: float64

Only 53 are above 100000, so let's cut those.

In [24]:
autos = autos[autos["price"] <= 100000]
autos["price"].describe()

count    48185.000000
mean      5796.099741
std       7525.532405
min        100.000000
25%       1250.000000
50%       3000.000000
75%       7499.000000
max      99900.000000
Name: price, dtype: float64

Next, let's clean up the various date columns in the data. First, we'll trim down the date_crawled, last_seen, and ad_created columns so that we're just looking at days.

In [25]:
autos["date_crawled"] = autos["date_crawled"].str[:10]
autos["ad_created"] = autos["ad_created"].str[:10]
autos["last_seen"] = autos["last_seen"].str[:10]

Now, let's take a look at each of these columns and see how the data stored within them is distributed.

In [26]:
autos["date_crawled"].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025340
2016-03-06    0.014050
2016-03-07    0.036090
2016-03-08    0.033164
2016-03-09    0.033019
2016-03-10    0.032313
2016-03-11    0.032624
2016-03-12    0.036920
2016-03-13    0.015690
2016-03-14    0.036692
2016-03-15    0.034305
2016-03-16    0.029470
2016-03-17    0.031504
2016-03-18    0.012867
2016-03-19    0.034762
2016-03-20    0.037813
2016-03-21    0.037190
2016-03-22    0.032811
2016-03-23    0.032292
2016-03-24    0.029449
2016-03-25    0.031504
2016-03-26    0.032292
2016-03-27    0.031109
2016-03-28    0.034949
2016-03-29    0.034139
2016-03-30    0.033703
2016-03-31    0.031856
2016-04-01    0.033662
2016-04-02    0.035633
2016-04-03    0.038601
2016-04-04    0.036567
2016-04-05    0.013054
2016-04-06    0.003175
2016-04-07    0.001390
Name: date_crawled, dtype: float64

In [27]:
autos["ad_created"].value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000042
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000042
2016-02-05    0.000042
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000042
2016-02-14    0.000042
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000042
2016-02-19    0.000062
2016-02-20    0.000042
2016-02-21    0.000062
                ...   
2016-03-09    0.033122
2016-03-10    0.032022
2016-03-11    0.032936
2016-03-12    0.036754
2016-03-13    0.017059
2016-03-14    0.035322
2016-03-15    0.034035
2016-03-16    0.029968
2016-03-17    0.031151
2016-03-18    0.013573
2016-03-19    0.033641
2016-03-20    0.037875
2016-03-21 

In [28]:
autos["last_seen"].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001079
2016-03-06    0.004317
2016-03-07    0.005437
2016-03-08    0.007326
2016-03-09    0.009567
2016-03-10    0.010646
2016-03-11    0.012411
2016-03-12    0.023804
2016-03-13    0.008882
2016-03-14    0.012639
2016-03-15    0.015876
2016-03-16    0.016437
2016-03-17    0.028121
2016-03-18    0.007305
2016-03-19    0.015773
2016-03-20    0.020650
2016-03-21    0.020546
2016-03-22    0.021376
2016-03-23    0.018574
2016-03-24    0.019736
2016-03-25    0.019114
2016-03-26    0.016623
2016-03-27    0.015544
2016-03-28    0.020836
2016-03-29    0.022310
2016-03-30    0.024717
2016-03-31    0.023846
2016-04-01    0.022870
2016-04-02    0.024883
2016-04-03    0.025132
2016-04-04    0.024551
2016-04-05    0.124935
2016-04-06    0.221999
2016-04-07    0.132137
Name: last_seen, dtype: float64

There doesn't seem to be anything too interesting above, except for the fact that the number of ads created on each day seemed to jump massively between February and March of 2016.

In [29]:
autos["registration_year"].describe()

count    48185.000000
mean      2004.730456
std         87.932039
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Moving on to data on registration year, we can clearly see some incorrect entries, such as the min and max values of 1000 and 9999 respectively. The various quartile values seem reasonable, however, as does the mean.

According to the wikipedia page on German vehicle registration plates, the first licence plates weren't required until 1906. Since we also know that a car can't be registered after the date on which it was listed on ebay, we can exclude any registration years after 2016, as well. So, let's get rid of these entries from our dataset.

In [30]:
autos = autos[autos["registration_year"].between(1906, 2016)]
autos["registration_year"].describe()

count    46313.000000
mean      2002.937750
std          7.118096
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [31]:
autos["registration_year"].value_counts(normalize = True).sort_index()

1910    0.000043
1927    0.000022
1929    0.000022
1931    0.000022
1934    0.000043
1937    0.000086
1938    0.000022
1939    0.000022
1941    0.000043
1943    0.000022
1948    0.000022
1950    0.000022
1951    0.000022
1952    0.000022
1953    0.000022
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000130
1960    0.000475
1961    0.000130
1962    0.000086
1963    0.000173
1964    0.000259
1965    0.000367
1966    0.000475
1967    0.000561
1968    0.000540
          ...   
1987    0.001555
1988    0.002850
1989    0.003692
1990    0.007169
1991    0.007298
1992    0.007903
1993    0.009069
1994    0.013517
1995    0.025738
1996    0.029257
1997    0.041543
1998    0.050461
1999    0.062164
2000    0.067001
2001    0.056744
2002    0.053484
2003    0.058148
2004    0.058277
2005    0.062855
2006    0.057608
2007    0.049058
2008    0.047676
2009    0.044912
2010    0.034180
2011    0.034915
2012    0.028178
2013    0.017209
2014    0.0142

It appears that most of the cars from our dataset were registered after 1990. To check, let's see what proportion of the cars in our dataset have been registered in or after 1999.

In [32]:
registered_after_1990 = autos.loc[autos["registration_year"] >= 1990, "registration_year"]

In [33]:
registered_after_1990.describe()

count    45042.000000
mean      2003.602660
std          5.751876
min       1990.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

As expected, the vast majority were registered in or after 1990, more than 90%, in fact.

Now, we'll take a look at the data on the car's brands. 

In [35]:
autos["brand"].value_counts(ascending = False, normalize = True).head(10)

volkswagen       0.211582
bmw              0.110207
opel             0.107335
mercedes_benz    0.096668
audi             0.086822
ford             0.069872
renault          0.047114
peugeot          0.029884
fiat             0.025630
seat             0.018267
Name: brand, dtype: float64

So the most popular brand is Volkswagen, followed by BMW, Opel, Mercedes Benz, and finally Audi. Let's aggregate the price data on some of these brands to find the mean price of a listing for each of the cars. We can combine boolean indexing with a loop here to aggregate this data from our dataset.

In [36]:
brands = ["volkswagen", "bmw", "opel", "mercedes_benz", "audi"]

brand_price_mean = {}

for b in brands:
    selected_rows = autos[autos["brand"] == b]
    mean_price = selected_rows["price"].mean()
    brand_price_mean[b] = mean_price
    
print(brand_price_mean)

{'volkswagen': 5436.950096948668, 'bmw': 8249.652429467085, 'opel': 3005.4960772480385, 'mercedes_benz': 8573.484922939468, 'audi': 9339.529967669734}


On average, BMWs, Mercedes Benzs, and Audis sell for the highest prices, with Audis selling for over 9000 on average. Comparatively, Volkswagens, at 5436, and Opels, at 3005, are much cheaper.

But, how do we know that the price differences we've just observed are not due to some other factor, such as the mileage of the cars?

Let's aggregate the average km travelled for each of these brands and see how the km travelled for the more expensive brands compares to those that are less expensive.

In [37]:
brand_km_mean = {}

for b in brands:
    selected_rows = autos[autos["brand"] == b]
    mean_km = selected_rows["odometer_km"].mean()
    brand_km_mean[b] = mean_km

print(brand_km_mean)
print(brand_price_mean)

{'volkswagen': 128799.87753852434, 'bmw': 132756.66144200627, 'opel': 129384.42969221485, 'mercedes_benz': 131088.89881617154, 'audi': 129276.29942800298}
{'volkswagen': 5436.950096948668, 'bmw': 8249.652429467085, 'opel': 3005.4960772480385, 'mercedes_benz': 8573.484922939468, 'audi': 9339.529967669734}


As it turns out, the mean km travelled for each of the brands is relatively similar, so the km travelled doesn't account for the difference in price.

This method for determining whether a factor affects the price is fine for now, but if we had more data it would be less easy to see. If we wanted to compare an additional factor as well, it would become even more difficult. We'd really like another way of visualizing our results so that we can more easily see whether any particular factor affects price. While we might want to do this in a graph later, for now we can use the series and dataframe constructors to group the results together into a single dataframe.

In [38]:
brand_price_mean_series = pd.Series(brand_price_mean)
brand_km_mean_series = pd.Series(brand_km_mean)
print(brand_price_mean_series)

volkswagen       5436.950097
bmw              8249.652429
opel             3005.496077
mercedes_benz    8573.484923
audi             9339.529968
dtype: float64


In [39]:
brand_dataframe = pd.DataFrame(brand_price_mean_series, columns = ["mean_price"])
brand_dataframe["mean_km_travelled"] = brand_km_mean_series
brand_dataframe

mean_price  mean_km_travelled
volkswagen     5436.950097      128799.877539
bmw            8249.652429      132756.661442
opel           3005.496077      129384.429692
mercedes_benz  8573.484923      131088.898816
audi           9339.529968      129276.299428

Now, if we wanted to see better whether the km travelled affected price, we could divide the first column by the second. If the km travelled was the sole factor influencing price, then we should get the same number for each calculation.

In [40]:
brand_price_over_km = {}

price = brand_dataframe.loc["volkswagen", "mean_price"]

for b in brands:
    price = brand_dataframe.loc[b, "mean_price"]
    km_travelled = brand_dataframe.loc[b, "mean_km_travelled"]
    brand_price_over_km[b] = price/km_travelled

brand_price_over_km_series = pd.Series(brand_price_over_km)    
brand_dataframe["price_over_km"] = brand_price_over_km_series

brand_dataframe

mean_price  mean_km_travelled  price_over_km
volkswagen     5436.950097      128799.877539       0.042212
bmw            8249.652429      132756.661442       0.062141
opel           3005.496077      129384.429692       0.023229
mercedes_benz  8573.484923      131088.898816       0.065402
audi           9339.529968      129276.299428       0.072245

Since the price_over_km varies between brands, we can conclude that the km travelled is not the determining factor for the brands price (although that does not mean it does not have some impact). If we wanted to see this, we'd need to control for the brand variable and calculate how much price changes with a change in km travelled. We could even see whether the relationship between km travelled and price is different for each brand. But we'll leave this project here for now.